pkg/apis/apiserver/project_unstructured_conversion.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2020 The Bulward Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiserver
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+
+	storagev1alpha1 "github.com/kubermatic/bulward/pkg/apis/storage/v1alpha1"
+)
+
+func ConvertToUnstructuredStorageV1Alpha1Project(project *Project, scheme *runtime.Scheme) (*unstructured.Unstructured, error) {
+	u := &unstructured.Unstructured{}
+	if _, err := chainConversion(scheme, project, &storagev1alpha1.Project{}, u); err != nil {
+		return nil, err
+	}
+	return u, nil
+}
+
+func ConvertFromUnstructuredStorageV1Alpha1Project(internalProjectv1alpha1 *unstructured.Unstructured, scheme *runtime.Scheme) (*Project, error) {
+	gvk, err := apiutil.GVKForObject(internalProjectv1alpha1, scheme)
+	if err != nil {
+		return nil, err
+	}
+	expectedGVK := storagev1alpha1.SchemeGroupVersion.WithKind("Project")
+	if gvk != expectedGVK {
+		return nil, fmt.Errorf("wrong GVK, expected %v, found %v", expectedGVK, gvk)
+	}
+	project := &Project{}
+	if _, err := chainConversion(scheme, internalProjectv1alpha1, &storagev1alpha1.Project{}, project); err != nil {
+		return nil, err
+	}
+	return project, nil
+}
+
+func ConvertToUnstructuredStorageV1Alpha1ProjectList(projects *ProjectList, scheme *runtime.Scheme, namespace string) (*unstructured.UnstructuredList, error) {
+	accesssor, err := meta.ListAccessor(projects)
+	if err != nil {
+		return nil, err
+	}
+	spl := &unstructured.UnstructuredList{}
+	for _, it := range projects.Items {
+		project, err := ConvertToUnstructuredStorageV1Alpha1Project(&it, scheme)
+		if err != nil {
+			return nil, err
+		}
+		spl.Items = append(spl.Items, *project)
+	}
+	spl.SetSelfLink(fmt.Sprintf("/apis/%s/%s/namespaces/%s/%s", storagev1alpha1.GroupVersion.Group, storagev1alpha1.GroupVersion.Version, namespace, internalProjectResource))
```

Comment:
Can you please add a comment here so we don't forget the reason that we put the `namespace` in the selflink?

Reply:
actually this is wrong. ill remove it

> > I generated the protobuf code for storage and apiserver, so now the tests are passing.
> > What is left to do:
> > 
> > * add protobuf generation to Makefile
> > * minor cleanup/code extraction of common code between project and organization
> 
> Is that ok for you that we add the protobuf generation command to our codegen.sh? We don't need to add anything to dev iamge, since it is preinstalled already.
> 
> For cleanup, we can do that in another PR.

Sure we can add it to the codegen. But we need to add `protoc` to the dev image as it is needed for the protobuf generation.

`apiserver/bin/go-to-protobuf --packages github.com/kubermatic/bulward/pkg/apis/apiserver/v1alpha1,github.com/kubermatic/bulward/pkg/apis/storage/v1alpha1 --apimachinery-packages -k8s.io/apimachinery/pkg/util/intstr,-k8s.io/apimachinery/pkg/api/resource,-k8s.io/apimachinery/pkg/runtime/schema,-k8s.io/apimachinery/pkg/runtime,-k8s.io/apimachinery/pkg/apis/meta/v1,-sigs.k8s.io/apiserver-builder-alpha/pkg/builders --drop-embedded-fields k8s.io/apimachinery/pkg/apis/meta/v1.TypeMeta,k8s.io/apimachinery/pkg/runtime.Serializer --proto-import=./vendor --vendor-output-base=./vendor/`

Reply:
Ok, can you please explain why this is wrong? I see we also set selflink for organization, if this is wrong, maybe we should also remove that

Reply:
Sorry, sure. I removed it already for both org and project.

The thing is that the selflik should be a functional link to a specific resource, like for example a specific org or project we can build it like `.../namespaces/test-org/test-project`. 

As I see it there is no such thing for lists, unless we want to use the listoptions and somehow generate the selfink. From what I saw in practice, the selflink for lists is empty.

Reply:
Ok, I see. Thanks!